### DISCUSSION TITLE
UMAP and HDBSCAN Tuning

### DISCUSSION THREAD

**Discussion by drob-xx** [MaartenGr/BERTopic#600] [General]
Moving convo here from #582 with @cedivad

> My latest insight is that I need n_neighbors to be a significant percentage of my UMAP training size (as it previously was set to just 150 on a 4M dataset and it's now 2000 on a 400k dataset). I'm following your insight above and trying to generate as many runs as possible to see if I can find the correct metaparameters for my job.

Ok. What is your goal? What wasn't working that tuning UMAP will fix?

> I could share my scripts? If the size of your collection is big enough the speedup over a raw BERTopic call is important, and iterating quickly is important as we are learning. But I'm guessing simplicity is king on smaller datasets.

For sure - I already don't use multiple runs of BERTopic for tuning - I break out the embeddings I need and run UMAP and HDBSCAN independently.

**Comment by drob-xx** [comment]
@cedivad, bumping this, would love to see your scripts and find out how your investigations are going. I've just started taking a closer look at UMAP settings, am also playing around with other reduction algos.

**Comment by cedivad** [comment]
Ouch. This is painful, because I spent way too long working on this and basically concluded nothing (I can barely see improvements).

For UMAP, I tried different `n_neighbors`, sample size as % of the full dataset and `min_distance`.

This generated a few different UMAP models, that I then used to train a bunch of different HDBSCAN models. I ranked the resulting models using a different validation dataset, and then collected stats such as uncategorized count, median category size, standard deviation of the category sizes (to understand if a handful of categories had all of the results, that is something I didn't want). 

cuML UMAP's `n_neighbors` seems to be topping out at ~2200, even with a 80GB GPU, but I ended up using a 5216k-big subsample of the original data, that has a terrible `n_neighbors` to data size ratio. I settled with having a 50% uncategorized count, and "fixed" it with calculating `hdbscan.membership_vector`, that is by far the slowest part of my pipeline, but a necessary evil, apparently.

Any distance but `min_dist` = 0 was apparently a terrible idea.

Here are the results and scripts:

https://gist.github.com/cedivad/ca09c3c10510c82961b4ca7a3e6d980d

**Reply by drob-xx** [reply]
Yah, alot of this work is trying out things that take a long time and don't work :(. Thanks very much for sharing. One thing that's helped me immensely are [parallel coordinates plots](https://plotly.com/python/parallel-coordinates-plot/) which turn a page full of numbers into something more understandable (at least for me).

I'm just working on some UMAP tuning right now, and I've done a little bit a couple of weeks ago. I don't see a huge effect. It changes things, but the overall impact on my models seems negligible. 

I have a bunch of questions about your (interesting) project, so let me know if you are open to sharing more specific info. I've basically been working with two datasets all year and while I'm happy so far with the progress I've made I have no idea how anything I've seen applies more generally.

Thanks for sharing!

Dan

**Reply by cedivad** [reply]
I'm happy to answer to everything I can, but I don't believe my answers would be of much use to an experienced user such as yourself. Your two ideas of segmenting the dataset and re-clustering the unassigned topics are beautiful!

**Reply by drob-xx** [reply]
I'm really interested in seeing more use cases to figure out if this approach is generally viable. As I've said I've basically been working with just two corpora and they are similar in character. I have no idea if anything I'm doing would positively impact your case, or just lead to a dead end. Having said that, if you are interested in talking through this then I'm curious about:

- What is your project - I see you are dealing with a lot of documents.
- What do you see as your current stumbling block. It sounds like you want to get as many documents positively labeled as possible? What is your goal?
- I'm curious about the probabilities output which it sounds like you are using for classification. I have avoided that route because it is so resource intensive. I've played extensively with probability distributions with LDA and Top2Vec, and found those weightings to be pretty useless. I really don't know much about how sensitive the HDBSCAN generated probabilities are.
- It sounds like you are generating a *ton* of topics and that you are using statistical measures to determine how 'good' they are. With so many topics what is the end-goal? What would you do with these topics once generated?
- Have you attempted to plot your data on a 2D TSNE reduction of the embeddings? I find that is a very useful way to get a birds eye view of the topic distributions. I haven't had a chance to play with the new BERTopic visualizations yet. In your case with so many documents I would guess that it would probably make a lot of sense to create a sample of your data to plot - but maybe 100-200K documents would be a large enough sample to accurately represent the entire corpus.

This is a lot - and I'm sure you are busy. If you are inclined I'm really curious, if this is out of scope I entirely understand!

**Comment by drob-xx** [comment]
@cedivad @MaartenGr,

This afternoon I was continuing to work on the UMAP/HDSBSCAN tuning. The 'problem' I'm trying to solve is that although I've found what seems to be good parameters for my corpus, when I visually examine the tsne 2D of the BERT mappings there are areas that seem like they were 'missed' - either they look like topic clusters of their own or they seem like they belong to clusters that have been labeled. When I examine individual documents within these groupings they still seem like they should be recognized.

Coming up with better parameters has made a big difference, but I am still vexed that it could be better than it is. On another dataset which I've talked about before I found that the clustering was greatly improved by segmenting the corpus. In that case it involved creating essentially 6 clusters and breaking one of them off and then working just with that grouping. In the first instance I was getting 5 well defined clusters and one big 'blob'. By segmenting out the blob and applying different parameters it clustered very nicely indeed. 

With my most recent corpus I went back to something I think I tried months and months ago (and didn't work) - taking all the -1 documents and re-clustering.  Initially, the results look very good. There are some very small clusters which I think are semantically defensible as clusters but more importantly there are some significant clusters which essentially address the areas I had concerns about. I am pretty sure the difference is was that when I 're-clustered' previously I was using the same parameters - so the results were the same.

In both the experiences I mention above what I've done is segment the corpus and then applied different HDBSCAN parameters to the embeddings. Long story short - it seems like it is possible - if not likely - to have a corpus for which one set of parameter simply won't work well (or optimally). I'd be interested in your feedback.

**Comment by denson** [comment]
I suspect that in many cases using parametric UMAP [docs](https://umap-learn.readthedocs.io/en/latest/parametric_umap.html) , [YouTube video](https://youtu.be/4Fbf9CQZE00) might result in substantial improvements.  However, I'm not sure how to accomplish this within BERTopic. My only solution is to use BERT, parametric UMAP and HDBSCAN separately in a pipeline. It does appear that it would be fairly easy to include an option within BERTopic.
